File: python/src/day4/main.py
from src.common.file_utils import get_path, read_raw
import re


def part_one(filename: str) -> int:
    lines = read_raw(get_path(__file__, filename))
    pports = lines.split("\n\n")
    valid = []
    for p in pports:
        valid_f = 0
        p = p.replace("\n", " ")
        fields = p.split()
        for f in fields:
            k, v = f.split(":")
            if k != "cid":
                valid_f += 1
        if valid_f >= 7:
            valid.append(p)
    return len(valid)


def part_two(filename: str) -> int:
    lines = read_raw(get_path(__file__, filename))
    pports = lines.split("\n\n")
    valid = []
    for p in pports:
        valid_f = 0
        p = p.replace("\n", " ")
        fields = p.split()
        for f in fields:
            k, v = f.split(":")
            if k in checks:
                if checks[k](v):
                    valid_f += checks[k](v)
        if valid_f >= 7:
            valid.append(p)
    return len(valid)


def check_byr(s):
    if not s.isdigit():
        return False
    if len(s) != 4:
        return False
    return 1920 <= int(s) <= 2002


def check_iyr(s):
    if not s.isdigit():
        return False
    if len(s) != 4:
        return False
    return 2010 <= int(s) <= 2020


def check_eyr(s):
    if not s.isdigit():
        return False
    if len(s) != 4:
        return False
    return 2020 <= int(s) <= 2030


def check_hgt(s):
    # units = [c for c in s if c.isalpha()]
    r = re.search(r'(\d+)([a-z]{2})', s)
    if not r:
        return False
    n, u = "", ""
    if r:
        n, u = r.groups()
    if u == "in":
        return 59 <= int(n) <= 76
    if u == "cm":
        return 150 <= int(n) <= 193
    return False


def check_hcl(s):
    if s[0] != "#":
        return False
    rest = s[1:]
    if len(rest) != 6:
        return False
    for c in rest:
        if not c.isdigit() and (ord("a") > ord(c) or ord("f") < ord(c)):
            return False
    return True


def check_ecl(s):
    return s in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]


def check_pid(s):
    return len(s) == 9 and s.isdigit()


def check_cid(s):
    return False


checks = {
    "byr": check_byr,
    "iyr": check_iyr,
    "eyr": check_eyr,
    "hgt": check_hgt,
    "hcl": check_hcl,
    "ecl": check_ecl,
    "pid": check_pid,
    "cid": check_cid,
}

if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day4/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	var validPassports int
	passports := createPassports(filename)

	for _, p := range passports {
		if p.validatePart1() {
			validPassports += 1
		}
	}

	return validPassports
}

func Part2(filename string) int {
	var validPassports int
	passports := createPassports(filename)

	for _, p := range passports {
		if p.validatePart2() {
			validPassports += 1
		}
	}
	return validPassports
}

type passport map[string]string

func createPassports(filename string) []passport {
	lines := input.ReadRaw(filename)
	data := strings.Split(lines, "\n\n")
	passports := []passport{}
	for _, d := range data {
		splitBySpaces := strings.ReplaceAll(d, "\n", " ")
		p := passport{}
		for _, f := range strings.Split(splitBySpaces, " ") {
			pair := strings.SplitN(f, ":", 2)
			if len(pair) == 2 {
				p[pair[0]] = pair[1]
			}
		}
		passports = append(passports, p)
	}
	return passports
}

func (p passport) validatePart1() bool {
	requiredFields := []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"}
	for _, required := range requiredFields {
		_, found := p[required]
		if !found {
			return false
		}
	}
	return true
}

func (p passport) validatePart2() bool {
	if !p.isValidByr() {
		return false
	}
	if !p.isValidIyr() {
		return false
	}
	if !p.isValidEyr() {
		return false
	}
	if !p.isValidHgt() {
		return false
	}
	if !p.isValidHcl() {
		return false
	}
	if !p.isValidEcl() {
		return false
	}
	if !p.isValidPid() {
		return false
	}
	return true
}

func (p passport) isValidByr() bool {
	if p["byr"] == "" {
		return false
	}
	n := toInt(p["byr"])
	return 1920 <= n && n <= 2002
}

func (p passport) isValidIyr() bool {
	if p["iyr"] == "" {
		return false
	}
	n := toInt(p["iyr"])
	return 2010 <= n && n <= 2020
}

func (p passport) isValidEyr() bool {
	if p["eyr"] == "" {
		return false
	}
	n := toInt(p["eyr"])
	return 2020 <= n && n <= 2030
}

func (p passport) isValidHgt() bool {
	if p["hgt"] == "" {
		return false
	}
	re := regexp.MustCompile(`(\d+)([a-z]+)`)
	parts := re.FindStringSubmatch(p["hgt"])
	if len(parts) != 3 {
		return false
	}
	n, units := toInt(parts[1]), parts[2]
	if units == "in" {
		return 59 <= n && 76 >= n
	}
	if units == "cm" {
		return 150 <= n && n <= 193
	}
	return false
}

func (p passport) isValidHcl() bool {
	if p["hcl"] == "" {
		return false
	}
	re := regexp.MustCompile(`#[a-f\d]{6}`)
	return re.MatchString(p["hcl"])
}

func (p passport) isValidEcl() bool {
	if p["ecl"] == "" {
		return false
	}
	cols := []string{"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}
	for _, c := range cols {
		if c == p["ecl"] {
			return true
		}
	}
	return false
}

func (p passport) isValidPid() bool {
	if p["pid"] == "" {
		return false
	}
	re := regexp.MustCompile(`\d{9}`)
	return re.MatchString(p["pid"])
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}
